File: DynamicProgramming_problems/CoinChangeBottomUp.java

// Problem link - https://leetcode.com/problems/coin-change/description/

package DynamicProgramming_problems;

public class CoinChangeBottomUp {
    public static int[] dp;
    public static int[] c;

    private static int coinChangeBU(int amtVal){
        dp = new int[amtVal+1];
        dp[0] = 0;

        for(int amount=1; amount<=amtVal; amount++){
            int ans = Integer.MAX_VALUE;
            int n = c.length;
            
            for(int i=0; i<n; i++){
                if(amount >= c[i]){
                    ans = Math.min(ans, dp[amount - c[i]]);
                }
            }

            // if ans is still infinity then coinChange is not possible
            if(ans == Integer.MAX_VALUE){
                dp[amount] = Integer.MAX_VALUE;
            }else{
                // ans + 1 --> here + 1 is for to react that way
                dp[amount] = 1 + ans;
            }
        }
        return dp[amtVal];
    }
    public static int coinChange(int[] coins, int amount) {
        c = coins;
        int minCoins = coinChangeBU(amount);
        return minCoins == Integer.MAX_VALUE ? -1 : minCoins;
    }

    public static void main(String[] args) {
        int[] coins = {1,2,5};
        int amount = 11;
        int ans = coinChange(coins, amount);
        System.out.println(ans);
    }
}

File: DynamicProgramming_problems/CoinChangeBottomUp.js

// Problem link - https://leetcode.com/problems/coin-change/description/

let dp;
let c;

function coinChangeBottomUpDp(amtVal){
    dp = Array(amtVal+1);
    dp[0] = 0;

    for(let amount=1; amount<=amtVal; amount++){
        let n = c.length;
        let res = Number.MAX_VALUE;

        // iterate over n number of coins
        for(let i=0; i<n; i++){
            if(amount >= c[i]){
                res = Math.min(res, dp[amount - c[i]]);
            }
        }

        // if ans is still Number.MAX_VALUE, then the given amount is not possible
        if(res === Number.MAX_VALUE){
            dp[amount] = Number.MAX_VALUE;
        }else{
            // otherwise calculate dp[amount]
            dp[amount] = 1 + res;
        }
    }
    return dp[amtVal];
}

let coinChange = function(coins, amount) {
    c = coins;
    let minCoins = coinChangeBottomUpDp(amount);
    return minCoins === Number.MAX_VALUE ? -1 : minCoins;
};

let coins = [1,2,5];
let amount = 11;
let ans = coinChange(coins, amount);
console.log(ans);